Distribution/Server/Features/Users.hs

```diff
@@ -468,6 +493,66 @@ userFeature  usersState adminsState
+           uid <- lookupUserName (UserName username)
+           guardAuthorised [IsUserId uid, InGroup adminGroup]
+           userInfo <- lookupUserInfo uid
+           let (UserTokenMap knownTokens) = userTokens userInfo
+           template <- getTemplate templates "manage.html"
+           let mkTok (t, desc) =
+                   templateDict
+                   [ templateVal "hash" (display t)
+                   , templateVal "description" desc
+                   ]
+           ok $ toResponse $ template
+               [ "username" $= username
+               , "tokens" $= map mkTok (Map.toList knownTokens)
+               ]
+
+    runUserManagementAction :: DynamicPath -> ServerPartE Response
+    runUserManagementAction dpath =
+        do (UserName username) <- userNameInPath dpath
+           uid <- lookupUserName (UserName username)
+           guardAuthorised [IsUserId uid, InGroup adminGroup]
+           action <- look "action"
+           case action of
+             "new-auth-key" ->
+                 do origTok <- liftIO generateOriginalToken
+                    let storeTok = convertToken origTok
+                    desc <- T.pack <$> look "description"
+                    res <- updateState usersState (AddAuthToken uid storeTok desc)
+                    template <- getTemplate templates "token-created.html"
+                    case res of
+                        Nothing ->
+                            ok $ toResponse $ template
+                            [ "username" $= username
+                            , "token" $= viewOriginalToken origTok
+                            ]
+                        Just Users.ErrNoSuchUserId ->
+                            errInternalError [MText "uid does not exist"]
+             "revoke-auth-key" ->
+                 do authToken <- parseAuthToken . T.pack <$> look "auth-key"
+                    case authToken of
+                      Left err ->
+                          fail $ "Bad auth token: " ++ err
+                      Right at ->
+                          do res <-
+                                 updateState usersState (RevokeAuthToken uid at)
+                             template <- getTemplate templates "token-revoked.html"
+                             case res of
+                               Nothing ->
+                                   ok $ toResponse $ template
+                                   [ "username" $= username
+                                   ]
+                               Just (Left Users.ErrNoSuchUserId) ->
+                                   errInternalError [MText "uid does not exist"]
+                               Just (Right Users.ErrTokenNotOwned) ->
+                                   errBadRequest "Token not owned"
+                                   [MText "Cannot revoke this token, wrong owner!"]
+             _ -> fail "Fatal error: missing implementation for this action"
```

Comment:
This will generate a 5xx internal server error. You should instead use an error response that says it's an invalid action. Same with other uses of fail above. You probably want `errBadRequest`, like you're using here in other places.